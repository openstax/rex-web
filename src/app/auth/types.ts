export interface User {
  firstName: string;
  lastName: string;
  uuid: string;
  isNotGdprLocation: boolean;
}

export interface State {
  user: User | undefined;
  established: boolean;
}
