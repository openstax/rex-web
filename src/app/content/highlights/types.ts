import { Highlight, HighlightColorEnum, HighlightsSummary, NewHighlight } from '@openstax/highlighter/dist/api';
import { ScrollTarget } from '../../navigation/types';
import { FiltersChange } from '../components/popUp/types';
import { LinkedArchiveTreeNode } from '../types';

export interface SummaryHighlights {
  [locationId: string]: {[pageId: string]: HighlightData[]};
}

export type OrderedSummaryHighlights = Array<{
  location: LinkedArchiveTreeNode,
  pages: Array<{
    pageId: string;
    highlights: HighlightData[];
  }>
}>;

export interface SummaryFilters {
  locationIds: string[];
  colors: HighlightColorEnum[];
}

export interface SummaryFiltersUpdate {
  locations?: FiltersChange<LinkedArchiveTreeNode>;
  colors?: FiltersChange<HighlightColorEnum>;
}

export type CountsPerSource = NonNullable<HighlightsSummary['countsPerSource']>;
export type HighlightColorCounts = CountsPerSource[string];

export type SummaryHighlightsPagination = null | {
  sourceIds: string[];
  page: number;
  perPage: number;
};

export interface State {
  currentPage: {
    pageId: string | null,
    highlights: null | HighlightData[];
    hasUnsavedHighlight: boolean;
    focused?: string;
  };
  summary: {
    open: boolean,
    pagination: SummaryHighlightsPagination,
    totalCountsPerPage: CountsPerSource | null;
    filters: SummaryFilters,
    loading: boolean;
    highlights: SummaryHighlights | null;
  };
}

export interface HighlightScrollTarget extends ScrollTarget {
  type: 'highlight';
  id: string;
}

// scopeId should not be nullable in the swagger
export type NewHighlightPayload = NewHighlight & {scopeId: string};
export type HighlightData = Highlight & {scopeId: string};
