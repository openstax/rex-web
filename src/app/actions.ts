import { createStandardAction } from 'typesafe-actions';

export const receivePageFocus = createStandardAction('app/receivePageFocus')<boolean>();
