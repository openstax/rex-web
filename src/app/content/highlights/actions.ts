import { Highlight, NewHighlight, UpdateHighlightRequest } from '@openstax/highlighter/dist/api';
import { createStandardAction } from 'typesafe-actions';
import { LocationFilters } from '../components/popUp/types';
import {
  CountsPerSource,
  HighlightData,
  SummaryFilters,
  SummaryFiltersUpdate,
  SummaryHighlights,
  SummaryHighlightsPagination,
} from './types';

export const focusHighlight = createStandardAction('Content/Highlights/focus')<string>();
export const clearFocusedHighlight = createStandardAction('Content/Highlights/clear')();
export const createHighlight = createStandardAction('Content/Highlights/create')<NewHighlight, {
  revertingAfterFailure?: boolean,
  locationFilterId: string,
  pageId: string,
}>();
export const requestDeleteHighlight = createStandardAction('Content/Highlights/requestDelete')<Highlight, {
  locationFilterId: string,
  pageId: string,
}>();
export const receiveDeleteHighlight = createStandardAction('Content/Highlights/receiveDelete')<Highlight, {
  revertingAfterFailure?: boolean,
  locationFilterId: string,
  pageId: string,
}>();
export const updateHighlight = createStandardAction('Content/Highlights/update')<UpdateHighlightRequest, {
  revertingAfterFailure?: boolean,
  preUpdateData: UpdateHighlightRequest,
  locationFilterId: string,
  pageId: string,
}>();
export const receiveHighlights = createStandardAction(
  'Content/Highlights/receive'
)<{highlights: HighlightData[], pageId: string}>();
export const setAnnotationChangesPending = createStandardAction('Content/Highlights/setAnnotationChangesPending')<
  boolean
>();

export const openMyHighlights = createStandardAction('Content/Highlights/Summary/open')<void>();
export const closeMyHighlights = createStandardAction('Content/Highlights/Summary/close')<void>();
export const initializeMyHighlightsSummary = createStandardAction('Content/Highlights/Summary/init')<void>();

export const printSummaryHighlights = createStandardAction('Content/Highlights/Summary/print')();
export const toggleSummaryHighlightsLoading = createStandardAction('Content/Highlights/Summary/loading')<boolean>();

export const loadMoreSummaryHighlights = createStandardAction('Content/Highlights/Summary/loadMore')();
export const setSummaryFilters = createStandardAction('Content/Highlights/Summary/setFilters')<
  Partial<SummaryFilters>
>();
export const updateSummaryFilters = createStandardAction('Content/Highlights/Summary/updateFilters')<
  Partial<SummaryFiltersUpdate>
>();
export const receiveSummaryHighlights = createStandardAction('Content/Highlights/Summary/receiveHighlights')<
  SummaryHighlights,
  {
    pagination: SummaryHighlightsPagination,
    filters?: SummaryFilters,
    isStillLoading?: boolean
  }
>();
export const receiveHighlightsTotalCounts = createStandardAction(
  'Content/receiveHighlightsTotalCounts'
)<CountsPerSource, LocationFilters>();
