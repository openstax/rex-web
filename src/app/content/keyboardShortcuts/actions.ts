import { createStandardAction } from 'typesafe-actions';

export const openKeyboardShortcutsMenu = createStandardAction('Content/KeyboardShortcuts/Menu/open')<void>();
export const closeKeyboardShortcutsMenu = createStandardAction('Content/KeyboardShortcuts/Menu/close')<void>();
