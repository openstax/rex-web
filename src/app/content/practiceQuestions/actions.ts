import { createStandardAction } from 'typesafe-actions';
import { LinkedArchiveTreeSection } from '../types';
import { PracticeQuestion, PracticeQuestionsSummary, SetPracticeAnswer } from './types';

export const setSelectedSection = createStandardAction(
  'Content/PracticeQuestions/setSelectedSection'
)<LinkedArchiveTreeSection | null>();
export const setQuestions = createStandardAction('Content/PracticeQuestions/setQuestions')<PracticeQuestion[]>();
export const nextQuestion = createStandardAction('Content/PracticeQuestions/nextQuestion')<void>();
export const setAnswer = createStandardAction('Content/PracticeQuestions/setAnswer')<SetPracticeAnswer>();
export const finishQuestions = createStandardAction('Content/PracticeQuestions/finishQuestions')<void>();

export const receivePracticeQuestionsSummary = createStandardAction(
  'Content/PracticeQuestions/Summary/receive'
)<PracticeQuestionsSummary>();

export const openPracticeQuestions = createStandardAction('Content/PracticeQuestions/open')<void>();
export const closePracticeQuestions = createStandardAction('Content/PracticeQuestions/close')<void>();
