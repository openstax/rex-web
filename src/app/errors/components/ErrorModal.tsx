import React from 'react';
import { FormattedMessage } from 'react-intl';
import { connect } from 'react-redux';
import styled from 'styled-components/macro';
import Button from '../../components/Button';
import { supportCenterLink } from '../../components/Footer';
import htmlMessage from '../../components/htmlMessage';
import Modal from '../../components/Modal';
import { Body, BodyHeading, Footer, modalPadding } from '../../components/Modal/styles';
import { AppState, Dispatch } from '../../types';
import { hideErrorDialog } from '../actions';
import * as select from '../selectors';
import ErrorIdList from './ErrorIdList';

// tslint:disable-next-line:variable-name
const BodyErrorText = styled.div`
  padding: ${modalPadding * 0.5}rem 0;
`;

// tslint:disable-next-line:variable-name
const BodyWithLink = htmlMessage('i18n:error:boundary:body', BodyErrorText);

interface PropTypes {
  show: boolean;
  error?: Error;
  stack: string[];
  clearError: () => void;
}

// tslint:disable-next-line:variable-name
const ErrorModal = ({ show, clearError, stack }: PropTypes) => {
  if (!show) { return null; }

  return (
    <Modal className='error-modal' heading='i18n:error:boundary:heading' onModalClose={clearError}>
      <Body>
        <FormattedMessage id='i18n:error:boundary:sub-heading'>
          {(msg) => <BodyHeading>{msg}</BodyHeading>}
        </FormattedMessage>
        <BodyWithLink values={{supportCenterLink}}/>
        <ErrorIdList ids={stack} />
      </Body>
      <Footer>
        <FormattedMessage id='i18n:error:boundary:action-btn-text'>
          {(msg) => <Button
            data-testid='clear-error'
            onClick={clearError}
            variant='primary'
            >{msg}
          </Button>}
        </FormattedMessage>
      </Footer>
    </Modal>
  );
};

export default connect(
  (state: AppState) => ({
    show: select.showErrorDialog(state),
    stack: select.getMessageIdStack(state),
  }),
  (dispatch: Dispatch) => ({
    clearError: () => dispatch(hideErrorDialog()),
  })
)(ErrorModal);
