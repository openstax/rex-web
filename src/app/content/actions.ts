import { createStandardAction } from 'typesafe-actions';
import * as highlightingActions from './highlights/actions';
import * as practiceQuestionsActions from './practiceQuestions/actions';
import * as searchActions from './search/actions';
import * as studyGuidesActions from './studyGuides/actions';
import { ArchivePage, Book, Params, State } from './types';

export const search = searchActions;
export const highlighting = highlightingActions;
export const studyGuides = studyGuidesActions;
export const practiceQuestions = practiceQuestionsActions;

export const openToc = createStandardAction('Content/openToc')<void>();
export const closeToc = createStandardAction('Content/closeToc')<void>();
export const resetToc = createStandardAction('Content/resetToc')<void>();

export const receiveBuyPrintConfig = createStandardAction('Content/receiveBuyPrintConfig')<State['buyPrint']>();

export const requestBook = createStandardAction('Content/requestBook')<Params['book']>();
export const receiveBook = createStandardAction('Content/receiveBook')<Book>();

export const requestPage = createStandardAction('Content/requestPage')<Params['page']>();
export const receivePage = createStandardAction('Content/receivePage')<
  ArchivePage & {references: State['references']}
>();
export const receivePageNotFoundId = createStandardAction('Content/receivePageNotFoundId')<string>();

export const openNudgeStudyTools = createStandardAction('Content/openNudgeStudyTools')();
export const closeNudgeStudyTools = createStandardAction('Content/closeNudgeStudyTools')();
