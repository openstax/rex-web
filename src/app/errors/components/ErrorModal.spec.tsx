import React from 'react';
import renderer from 'react-test-renderer';
import createTestStore from '../../../test/createTestStore';
import TestContainer from '../../../test/TestContainer';
import { Store } from '../../types';
import { hideErrorDialog, recordSentryMessage } from '../actions';

import ErrorModal from './ErrorModal';

describe('ErrorModal', () => {
  let store: Store;
  let error: Error;
  let dispatch: jest.SpyInstance;

  beforeEach(() => {
    error = new Error('unknown error');
    store = createTestStore({ errors: { showDialog: true, error, sentryMessageIdStack: [] } });
    dispatch = jest.spyOn(store, 'dispatch');
  });

  it('matches snapshot', () => {
    const tree = renderer
      .create(<TestContainer store={store}><ErrorModal /></TestContainer>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshots with recorded error ids', () => {
    store.dispatch(recordSentryMessage('some-error-id'));
    const tree = renderer
      .create(<TestContainer store={store}><ErrorModal /></TestContainer>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('clears errors', () => {
    const tree = renderer.create(<TestContainer store={store}><ErrorModal /></TestContainer>);

    const btn = tree.root.findByProps({ 'data-testid': 'clear-error' });
    renderer.act(() => { btn.props.onClick(); });
    expect(dispatch).toHaveBeenCalledWith(hideErrorDialog());
  });
});
