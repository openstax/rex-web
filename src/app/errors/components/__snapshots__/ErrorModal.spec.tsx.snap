// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorModal matches snapshot 1`] = `
<div
  className="error-modal styles__ModalWrapper-sc-1gxqzoa-6 bfHzdf"
>
  <div
    className="styles__CardWrapper-sc-1gxqzoa-7 llgsuk"
  >
    <div
      className="styles__Card-sc-1gxqzoa-0 cPBKEa"
    >
      <header
        className="styles__Header-sc-1gxqzoa-1 knqvba"
      >
        <h1
          className="styles__Heading-sc-1gxqzoa-2 eqDSaw"
        >
          Error
        </h1>
        <svg
          aria-hidden="true"
          className="styles__CloseModalIcon-sc-1gxqzoa-8 cjHFEZ"
          focusable="false"
          height="15px"
          onClick={[Function]}
          version="1.1"
          viewBox="0 0 15 15"
          width="15px"
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="none"
            strokeWidth="1"
          >
            <g
              fill="currentColor"
              transform="translate(-302.000000, -18.000000)"
            >
              <g
                transform="translate(302.000000, 18.000000)"
              >
                <path
                  d="M7.5,5.41522791 L12.0331524,0.579865364 C12.3077536,0.286957429 12.7165503,0.24816296 12.946282,0.493210121 L13.9861449,1.60239723 C14.2158766,1.84744439 14.1795068,2.28349422 13.9049056,2.57640216 L9.37175324,7.41176471 L13.9049056,12.2471273 C14.1795068,12.5400352 14.2158766,12.976085 13.9861449,13.2211322 L12.946282,14.3303193 C12.7165503,14.5753665 12.3077536,14.536572 12.0331524,14.243664 L7.5,9.4083015 L2.96684761,14.243664 C2.69224642,14.536572 2.2834497,14.5753665 2.05371799,14.3303193 L1.01385508,13.2211322 C0.784123363,12.976085 0.820493178,12.5400352 1.09509437,12.2471273 L5.62824676,7.41176471 L1.09509437,2.57640216 C0.820493178,2.28349422 0.784123363,1.84744439 1.01385508,1.60239723 L2.05371799,0.493210121 C2.2834497,0.24816296 2.69224642,0.286957429 2.96684761,0.579865364 L7.5,5.41522791 Z"
                />
              </g>
            </g>
          </g>
        </svg>
      </header>
      <div
        className="styles__Body-sc-1gxqzoa-4 eIyaEL"
      >
        <h3
          className="styles__BodyHeading-sc-1gxqzoa-3 krnhWx"
        >
          Uh-oh, there's been a glitch
        </h3>
        <div
          className="ErrorModal__BodyErrorText-sc-1q7eiy-0 glzJJQ"
          dangerouslySetInnerHTML={
            Object {
              "__html": "
  We're not quite sure what went wrong. Refresh the page. If this doesn't solve
  the problem, visit our <a href=\\"https://openstax.secure.force.com/help\\" target=\\"_blank\\">Support Center</a>.
",
            }
          }
        />
      </div>
      <div
        className="styles__Footer-sc-1gxqzoa-9 eKsEVO"
      >
        <button
          className="Button-ayg7nk-0 bBYlAl"
          data-testid="clear-error"
          onClick={[Function]}
        >
          OK
        </button>
      </div>
    </div>
  </div>
  <div
    className="styles__Mask-sc-1gxqzoa-5 bsZwaD"
  />
</div>
`;

exports[`ErrorModal matches snapshots with recorded error ids 1`] = `
<div
  className="error-modal styles__ModalWrapper-sc-1gxqzoa-6 bfHzdf"
>
  <div
    className="styles__CardWrapper-sc-1gxqzoa-7 llgsuk"
  >
    <div
      className="styles__Card-sc-1gxqzoa-0 cPBKEa"
    >
      <header
        className="styles__Header-sc-1gxqzoa-1 knqvba"
      >
        <h1
          className="styles__Heading-sc-1gxqzoa-2 eqDSaw"
        >
          Error
        </h1>
        <svg
          aria-hidden="true"
          className="styles__CloseModalIcon-sc-1gxqzoa-8 cjHFEZ"
          focusable="false"
          height="15px"
          onClick={[Function]}
          version="1.1"
          viewBox="0 0 15 15"
          width="15px"
        >
          <g
            fill="none"
            fillRule="evenodd"
            stroke="none"
            strokeWidth="1"
          >
            <g
              fill="currentColor"
              transform="translate(-302.000000, -18.000000)"
            >
              <g
                transform="translate(302.000000, 18.000000)"
              >
                <path
                  d="M7.5,5.41522791 L12.0331524,0.579865364 C12.3077536,0.286957429 12.7165503,0.24816296 12.946282,0.493210121 L13.9861449,1.60239723 C14.2158766,1.84744439 14.1795068,2.28349422 13.9049056,2.57640216 L9.37175324,7.41176471 L13.9049056,12.2471273 C14.1795068,12.5400352 14.2158766,12.976085 13.9861449,13.2211322 L12.946282,14.3303193 C12.7165503,14.5753665 12.3077536,14.536572 12.0331524,14.243664 L7.5,9.4083015 L2.96684761,14.243664 C2.69224642,14.536572 2.2834497,14.5753665 2.05371799,14.3303193 L1.01385508,13.2211322 C0.784123363,12.976085 0.820493178,12.5400352 1.09509437,12.2471273 L5.62824676,7.41176471 L1.09509437,2.57640216 C0.820493178,2.28349422 0.784123363,1.84744439 1.01385508,1.60239723 L2.05371799,0.493210121 C2.2834497,0.24816296 2.69224642,0.286957429 2.96684761,0.579865364 L7.5,5.41522791 Z"
                />
              </g>
            </g>
          </g>
        </svg>
      </header>
      <div
        className="styles__Body-sc-1gxqzoa-4 eIyaEL"
      >
        <h3
          className="styles__BodyHeading-sc-1gxqzoa-3 krnhWx"
        >
          Uh-oh, there's been a glitch
        </h3>
        <div
          className="ErrorModal__BodyErrorText-sc-1q7eiy-0 glzJJQ"
          dangerouslySetInnerHTML={
            Object {
              "__html": "
  We're not quite sure what went wrong. Refresh the page. If this doesn't solve
  the problem, visit our <a href=\\"https://openstax.secure.force.com/help\\" target=\\"_blank\\">Support Center</a>.
",
            }
          }
        />
        <div
          className="ErrorIdList-sc-1coujqm-0 dXzSc"
        >
          some-error-id
        </div>
      </div>
      <div
        className="styles__Footer-sc-1gxqzoa-9 eKsEVO"
      >
        <button
          className="Button-ayg7nk-0 bBYlAl"
          data-testid="clear-error"
          onClick={[Function]}
        >
          OK
        </button>
      </div>
    </div>
  </div>
  <div
    className="styles__Mask-sc-1gxqzoa-5 bsZwaD"
  />
</div>
`;
